src/persist-client/src/operators/shard_source.rs

```diff
@@ -0,0 +1,467 @@
+        mz_ore::task::spawn(|| "LeaseReturner", async move {
+            while let Some(leased_part) = consumed_part_rx.recv().await {
+                subscription
+                    .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
+            }
+        });
+    });
+
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("shard_source_descs({})", name), scope.clone());
+    let mut flow_control_input = builder.new_input(flow_control_input, Pipeline);
+    let (mut descs_output, descs_stream) = builder.new_output();
+
+    let shutdown_button = builder.build(move |caps| async move {
+        let mut cap_set = CapabilitySet::from_elem(caps.into_element());
+
+        let mut current_ts = timely::progress::Timestamp::minimum();
+        let mut inflight_bytes = 0;
+        let mut inflight_parts: Vec<(Antichain<T>, usize)> = Vec::new();
+
+        loop {
+            // While we have budget left for fetching more parts, read from the
+            // subscription and pass them on.
+            while inflight_bytes < flow_control_max_inflight_bytes {
+                match pinned_stream.next().await {
+                    Some(Ok((parts, progress, size_in_bytes))) => {
+                        let session_cap = cap_set.delayed(&current_ts);
+                        let mut descs_output = descs_output.activate();
+                        let mut descs_session = descs_output.session(&session_cap);
+
+                        if size_in_bytes > 0 {
+                            inflight_parts.push((progress.clone(), size_in_bytes));
+                            inflight_bytes += size_in_bytes;
+                            trace!(
+                                "shard {} putting {} bytes inflight. total: {}. batch frontier {:?}",
+                                shard_id,
+                                size_in_bytes,
+                                inflight_bytes,
+                                progress,
+                            );
+                        }
+
+                        for part_desc in parts {
+                            // Give the part to a random worker.
+                            let worker_idx = usize::cast_from(Instant::now().hashed()) % num_workers;
+                            descs_session.give((worker_idx, part_desc.into_exchangeable_part()));
+                        }
+
+                        cap_set.downgrade(progress.iter());
+                        match progress.elements() {
+                            [ts] => {
+                                current_ts = ts.clone();
+                            }
+                            [] => {
+                                cap_set.downgrade(&[]);
+                                return;
+                            }
+                            _ => panic!("TODO: support partially ordered timestamps in shard_source"),
```

Comment:
That TODO is pretty hard to fix. DD goes to great lengths to split batches up because timely doesn't support this but it's unclear how we could do the same kind of splitting with persist batches even if timely got the feature.

https://github.com/TimelyDataflow/differential-dataflow/blob/master/src/operators/arrange/arrangement.rs#L572-L575

Given that this isn't going to get fixed any time soon can we put a `TotalOrder` bound until then instead of testing at runtime? With that bound this method becomes available:

```rust
match progress.as_option() {
    Some(ts) => current_ts = ts.clone(),
    None => { .. }
}
```